includes/core/classes/class-calendars.php

```diff
@@ -0,0 +1,827 @@
+
+		$charset = strtolower( get_option( 'blog_charset' ) );
+
+		header( 'Content-Description: File Transfer' );
+
+		// Ensure proper content type for the calendar file.
+		header( 'Content-Type: text/calendar; charset=' . $charset );
+
+		// Force download in most browsers.
+		header( 'Content-Disposition: attachment; filename="' . $filename . '"' );
+
+		// Avoid browser caching issues.
+		header( 'Cache-Control: no-store, no-cache, must-revalidate' );
+		header( 'Cache-Control: post-check=0, pre-check=0', false );
+		header( 'Pragma: no-cache' );
+		header( 'Expires: 0' );
+
+		// Prevent content sniffing which might lead to MIME type mismatch.
+		header( 'X-Content-Type-Options: nosniff' );
+	}
+
+	/**
+	 * Output the event(s) as an iCalendar (.ics) file.
+	 *
+	 * @return void
+	 */
+	public static function send_ics_file() {
+		// Start output buffering to capture all output.
+		ob_start();
+
+		// Prepare the filename.
+		$filename = self::generate_ics_filename();
+
+		// Send headers for downloading the .ics file.
+		self::send_ics_headers( $filename );
+
+		// Output the generated iCalendar content.
+		$get_ical_method = ( is_feed() ) ? 'get_ical_feed' : 'get_ical_file';
+		echo wp_kses_post( self::$get_ical_method() );
+
+		// Get the generated output and calculate file size.
+		$ics_content = ob_get_contents();
+		$filesize    = strlen( $ics_content );
+
+		// Send the file size in the header.
+		header( 'Content-Length: ' . $filesize );
+
+		// End output buffering and clean up.
+		ob_end_clean();
+
+		// Output the iCalendar content.
+		echo wp_kses_post( $ics_content );
+
+		exit(); // Terminate the script after the file has been output.
+	}
+
+	/**
+	 * @author Stephen Harris (@stephenharris)
+	 * @source https://github.com/stephenharris/Event-Organiser/blob/develop/includes/event-organiser-utility-functions.php#L1663
+	 *
+	 * Fold text as per [iCal specifications](http://www.ietf.org/rfc/rfc2445.txt)
+	 *
+	 * Lines of text SHOULD NOT be longer than 75 octets, excluding the line
+	 * break. Long content lines SHOULD be split into a multiple line
+	 * representations using a line "folding" technique. That is, a long
+	 * line can be split between any two characters by inserting a CRLF
+	 * immediately followed by a single linear white space character (i.e.,
+	 * SPACE, US-ASCII decimal 32 or HTAB, US-ASCII decimal 9). Any sequence
+	 * of CRLF followed immediately by a single linear white space character
+	 * is ignored (i.e., removed) when processing the content type.
+	 *
+	 * @ignore
+	 * @since 2.7
+	 * @param string $text The string to be escaped.
+	 * @return string The escaped string.
+	 */
+	private static function eventorganiser_fold_ical_text( string $text ): string {
+
+		$text_arr = array();
+
+		$lines = ceil( mb_strlen( $text ) / 75 );
+
+		for ( $i = 0; $i < $lines; $i++ ) {
+			$text_arr[ $i ] = mb_substr( $text, $i * 75, 75 );
+		}
+
+		return join( "\r\n ", $text_arr );
+	}
```

Comment:
I brazenly stole this code, because I knew it is working for years and I also knew that I couldn't do better.
But I don't just want to pirate Stephen Harris' code, without proper attribution - or rewriting it!